### PR TITLE
Better formatting of markers when a node spans multiple lines

### PR DESCRIPTION
**Pull request by aroberge**
This took me a long time to figure out how do do as I was working "downstream" (at the formatting stage) based on my experience with my own code.  In the end, it turns out to be a relatively small change (compared with my previous attempts).

I did a few tests (with ^ markers only) and everything seemed to work correctly.
Here's an example.

Before
```
       4 | def bad(i):
       5 |     x = [[[0]]]
-->    6 |     return x[0][0][0] + x[0
                                   ^^^
       7 |               ][
           ^^^^^^^^^^^^^^^^
       8 |                 i][
           ^^^^^^^^^^^^^^^^^^
       9 |                   0
      10 |               ]
```
After
```
       4 | def bad(i):
       5 |     x = [[[0]]]
-->    6 |     return x[0][0][0] + x[0
                                   ^^^
       7 |               ][
                         ^^
       8 |                 i][
                         ^^^^
       9 |                   0
      10 |               ]
```

And a different one, when the indentation of the markers in the first line is the smallest:
```
       4 | def bad(i):
       5 |     x = [[[0]]]
-->    6 |     return x[0
                      ^^^
       7 |               ][
                      ^^^^^
       8 |                 i][
                      ^^^^^^^
       9 |                   0
      10 |               ]
```

### CODE REVIEW

**Comment by aroberge** [comment]
I just realized that I had not run the tests. There are two tests failing, which should not be unexpected. However, one of them (with a line gap) likely indicate that the proposed change is not entirely correct.

File: stack_data/core.py [hunk 1]
```diff
@@ -669,6 +673,23 @@ def lines(self) -> List[Union[Line, LineGap]]:
...
+        indents = []
+        for i, piece in enumerate(pieces):
+            lines = [
+                Line(self, i)
+                for i in piece
+            ]  # type: List[Line]
+            for line in lines:
+                for line_range in line.executing_node_ranges:
+                    if line.is_current:
+                        indents.append(line_range.start)
```
Comment (alexmojaki) [inline stack_data/core.py:688]: This doesn't look right, rather than `is_current` I think it should be about whether it's the first line in the executing node, i.e. the only line that might have non-whitespace before the executing node starts.

**Comment by alexmojaki** [comment]
Yes, I think it'd be better to follow the pattern used for `leading_indent` to not include line gaps in the calculation. By the way, you should probably choose a more distinct name to avoid confusion, and add a leading underscore since I don't think users should touch this attribute.

Also, the computation is based on the executing node, but `range_from_node` is more general than that. I suggest adding a private cached property to FrameInfo and using it from `Line.executing_node_ranges`.

File: stack_data/core.py (hunk 1 above)
Comment (aroberge) [reply stack_data/core.py:688]: If you look at the second example above, the alignment is done with `x[0` which, I think looks better than if it were with `x=...` on the previous line.

**Comment by aroberge** [comment]
> Yes, I think it'd be better to follow the pattern used for leading_indent to not include line gaps in the calculation. By the way, you should probably choose a more distinct name to avoid confusion, and add a leading underscore since I don't think users should touch this attribute.

* No problem about choosing a different name.  I saw my code more as a proof-of-concept, definitely subject to changes.
* I'll have to take a look later about what you mention about leading_indent.

> Also, the computation is based on the executing node, but range_from_node is more general than that. I suggest adding a private cached property to FrameInfo and using it from Line.executing_node_ranges.

I'll have to look at the code to see what you mean. I may not have time to do so until Friday.

File: stack_data/core.py (hunk 1 above)
Comment (aroberge) [reply stack_data/core.py:688]: The idea I had was to minimize the amount of white space being highlighted, while keeping a vertical alignment of all the highlighted code.

Comment (alexmojaki) [reply stack_data/core.py:688]: Yes, but `x[0` is the start of the executing node in that case so the example wouldn't change. Basically your examples always have the current line being the same as the first highlighted line.

Comment (aroberge) [reply stack_data/core.py:688]: Isn't this always the case?

Comment (alexmojaki) [reply stack_data/core.py:688]:
```python
import stack_data
stack_data.Formatter().set_hook()

lst = []
print(lst
      .pop())
```

In 3.9:

```
       1 | import stack_data
       2 | stack_data.Formatter().set_hook()
       4 | lst = []
-->    5 | print(lst
                 ^^^
       6 |       .pop())
           ^^^^^^^^^^^^
```

But in 3.10:

```
       1 | import stack_data
       2 | stack_data.Formatter().set_hook()
       4 | lst = []
       5 | print(lst
                 ^^^
-->    6 |       .pop())
           ^^^^^^^^^^^^
```

https://bugs.python.org/issue39316

Comment (aroberge) [reply stack_data/core.py:688]: Ok, I was able to modify it so that I get the following for Python 3.10.
```
       1 | import stack_data
       2 | stack_data.Formatter().set_hook()
       4 | lst= []
       5 | print(lst
                 ^^^
-->    6 |       .pop())
                 ^^^^^^
```
and this for 3.9
```
       1 | import stack_data
       2 | stack_data.Formatter().set_hook()
       4 | lst= []
-->    5 | print(lst
                 ^^^
       6 |       .pop())
                 ^^^^^^
I```
A couple more examples for good measure:
```
-->    5 | print(lst
                 ^^^
       6 |          .pop())
                 ^^^^^^^^^
```
```
-->    5 | print(lst
                 ^^^
       6 |    .pop())
              ^^^^^^
```
Other than the indicator for the current line, I get the same markers for the indentation for both 3.9 and 3.10.

I will need to do more testing.

Unrelated: I've never attempted to do a multi-commit PR  (and I didn't create a separate branch). If I do another commit in my fork followed by a PR, will it be smart enough to add it to this one?
(I won't be able to do significantly more work on this right now.)

Comment (aroberge) [reply stack_data/core.py:688]: A final one for tonight:
```
       4 | print([1, 2,
                 ^^^^^^
       5 |     3, 4, 5]
               ^^^^^^^^
-->    6 |          .pop(8))
               ^^^^^^^^^^^^
```

Comment (aroberge) [reply stack_data/core.py]: It should be all ready now. 

Other than refactoring, I had to make an additional change so that using it to use markers for frames directly (as opposed as arising from a traceback), like you did in some tests with `<var>...</var>`, was broken by my code. Everything should now be fixed.

**Comment by alexmojaki** [comment]
> Also, the computation is based on the executing node, but range_from_node is more general than that

To clarify what I mean by this, change `result = print_stack()` in `tests/samples/example.py` to span multiple lines and see how it affects the `<var>` markers.

**Comment by aroberge** [comment]
> To clarify what I mean by this, change result = print_stack() in tests/samples/example.py to span multiple lines and see how it affects the <var> markers.

I'm sorry, but I do not understand what you mean by this, and thus how to do it.

**Comment by alexmojaki** [comment]
> Other than refactoring, I had to make an additional change so that using it to use markers for frames directly (as opposed as arising from a traceback), like you did in some tests with `<var>...</var>`, was broken by my code. Everything should now be fixed.

The problem had nothing to do with tracebacks or not, just the fact that `range_from_node` isn't always referring to the executing node. The use case was for the `stackprinter` library which highlighted specific variables rather than syntax highlighting.

Change

```python
result = print_stack()
```

to

```python
result = print_stack(
             )
```

and you should see a test failure. This is because you didn't fix the problem, you just avoided it by checking the number of lines.

**Comment by aroberge** [comment]
Ah, I think I finally understand what you mean. Sorry for being obtuse.

**Comment by aroberge** [comment]
All tests pass with this new version.

Changing `print_stack()` to insert a newline character.  Excluding some failures simply due to different line numbers, unmodified stack_data test fails with:
```
E         -   20 >     result = <exec>print_stack()</exec>
E         ?                                       -
E         +   20 >     result = <exec>print_stack(</exec>
E         +   21 | <exec>        )</exec>
```
New version fails with:
E         -   20 >     result = <exec>print_stack()</exec>
E         ?                                       -
E         +   20 >     result = <exec>print_stack(</exec>
E         +   21 |         <exec>)</exec>
```
The only difference is on the last line with the location of `<exec>`.

**Comment by alexmojaki** [comment]
The failed test should print out the actual output which you can copy and paste into the expected output.

File: stack_data/core.py [hunk 2]
```diff
@@ -669,6 +700,7 @@ def lines(self) -> List[Union[Line, LineGap]]:
         if not pieces:
             return []
 
+        common_markers_block_indent = self._markers_block_indent()
```
Comment (alexmojaki) [inline stack_data/core.py:703]: > not everyone using `lines` is interested in highlighting the executing node, hence my request to only do this computation in executing_node_ranges.

> I suggest adding a private cached property to FrameInfo and using it from Line.executing_node_ranges.

Comment (aroberge) [reply stack_data/core.py:703]: When I try to do this, I get a RecursionError.

As I understand it, here's why:

FrameInfo._markers_block_indent() creates **temporary** Line objects for its internal computation; these are not the cached FrameInfo.lines; those are properly obtained only after FrameInfo._markers_block_indent() has been calculated. 

I do need these temporary Line objects to figure out what common indentation should be used.

These temporary Line objects are created before FrameInfo._markers_block_indent() has finished. If I move the calculation in Line.executing_node_range, each temporary Line object will in turn call Frame.frame_info._markers_block_indent, which will create new temporary Line objects, etc.

Comment (alexmojaki) [reply stack_data/core.py:703]: You don't need temporary line objects - see how `leading_indent` is computed. But that's not really the point. Set the default `Line._markers_block_indent` to `None`, and check for that to decide if it needs to be computed.

Comment (aroberge) [reply stack_data/core.py:703]: This common marker_block_indent is used to compute Line.range_from_node; unlike Line.leading_indent, this is a NamedTuple (RangeInLine) which cannot be modified once computed; so, once a Line object has been created, this property can no longer be modified.  Even if it could, it still would not really help me, since an other relevant cached property, Line.executing_node_ranges, could not be changed.

I cannot compute a marker_block_indent based on a single Line; I need a block of lines to do so. 

So, I create a temporary version of all the lines that belong to an executing piece of a frame, using the default value for range.start for each line.  Each of these temporary lines has a (possibly incorrect) value of its cached property Line.executing_node_ranges, since I do not know the common indentation value to use at this point.

Then, I use these temporary Line objects to compute the "true" lines that belong to the FrameInfo, using this commmon indentation value.

In summary: Lines are, for practical purpose, immutable objects. To compute their common indentation, I need to use many of them (belonging to a common indented block), to determine their common indentation value. Since I cannot mutate these Line objects, I must recreate some new ones with the appropriate values.

To avoid creating temporary lines, at the very least, I would have to replace the NamedTuple RangeInLine by a mutable class, so that its `start` attribute could be changed.  Even then,

**Comment by alexmojaki** [comment]
Looks good! Sorry I was a bit blunt at times, I appreciate the contribution.

I'm merging now but I'm going to check that this doesn't inadvertently break anything (like IPython tests) before releasing.

If you're interested in contributing more, some other things to consider:

1. This PR doesn't affect the highlighting with pygments used by IPython and futurecoder.
2. You mentioned you preferred to include the blank lines which stack_data omits from its pieces. You could add an option to include them.
3. #21

**Comment by aroberge** [comment]
> 1. This PR doesn't affect the highlighting with pygments used by IPython and futurecoder.

I hadn't checked. I might try to do it.

> 2. You mentioned you preferred to include the blank lines which stack_data omits from its pieces. You could add an option to include them.

I had a brief look at that to see if I could do it (I would have first created a new issue before adding any new options.)  However, I didn't easily see how to do this; I guess I don't understand the code enough -- yet. Furthermore, I was thinking that I would like to add single empty lines skipped showing the line number, but adding a different marker than the one used by LineGap when skipping multiple lines, perhaps something like `!` as in

```
    1: print('hello')
     |
 -->6: raise Exception()
```
I might do something like this at the formatting stage in friendly-traceback first, to see if I like it enough after working with some real examples.

I have enough to do right now with friendly-traceback (and I still have to contribute translations to futurecoder) that tackling #21 is a very low priority for me right now.

**Comment by alexmojaki** [comment]
> However, I didn't easily see how to do this; I guess I don't understand the code enough -- yet.

I think it would be done by not filtering blank lines in `Source._clean_pieces`, followed by conditionally filtering them in `lines`.

> I still have to contribute translations to futurecoder

That would also be awesome! I don't know if you saw that there's now a fully translated Spanish version at https://es-latam.futurecoder.io/. It's being tested and polished in preparation for a wider release. I'm writing up some docs on how to contribute, I'll send it to you when it's ready.

**Comment by alexmojaki** [comment]
I'd like to release a new version soon with a fix for #21 so I suggest putting an upper bound on the stack_data version until you're ready.

**Comment by aroberge** [comment]
Fixed the stack_data version number for friendly-traceback.

I will test it with the new version as soon as I can after its release.

**Comment by alexmojaki** [comment]
I've released.

**Comment by aroberge** [comment]
Friendly-traceback runs unchanged with the latest stack_data release.  I suspect that I can now remove some code, but that will have to wait.

Friendly-traceback now requires stack_data >= 0.2.0.
